### ISSUE TITLE
Pubsub: Subscriber spawns one goroutine per message outstanding, causing OOM with many small messages.

### ISSUE THREAD

**Issue by dpcollins-google** [googleapis/google-cloud-go#1318]
## Client

PubSub

## Describe Your Environment

Any

Specifically, GCE Ubuntu n1-standard-16

## Expected Behavior

Pubsub subscriber handles 25GB of 1KB messages outstanding on a machine with 60GB of memory.

## Actual Behavior

Pubsub subscriber crashes due to memory limits and prints millions of lines of stacktrace to stderr.

This is due to the fact that at https://github.com/googleapis/google-cloud-go/blob/f23c43891e43fa5323eb751293c177f0a4196b1a/pubsub/subscription.go#L573 the client spawns a goroutine for every message.  Given that these have nonzero stack size, they can quickly overwhelm the memory limitation when generated in a tight loop due to receiving a new batch message.

## Proposed resolution

Replace the logic on line 573 linked above with an unbounded pool of goroutines.  This way, if the subscriber client is efficient in processing messages, the number of goroutines will remain small.  If the subscriber client is inefficient and all messages cause a block, the behavior will be identical to the current behavior.

**Comment by jba** [comment]
You can limit the number of active goroutines by using the MaxOutstandingMessages and MaxOutstandingBytes ReceiveSettings.

**Comment by dpcollins-google** [comment]
> You can limit the number of active goroutines by using the MaxOutstandingMessages and MaxOutstandingBytes ReceiveSettings.

This is not sufficient.  I'm subscribing at > 500MB per second with 1KB sized messages.  To keep subscribers saturated at all times, I'm trying to keep 25GB of work outstanding.  This means 25 million goroutines outstanding at once, which in aggregate consumes more memory than the messages I have outstanding.

**Comment by dpcollins-google** [comment]
Here's a heap profile from <10 seconds before the go runtime ran out of memory, showing 25GB of userland usage.  On a 60GB machine running nothing else but the OS, this should not cause problems.  
![profile001](https://user-images.githubusercontent.com/40498610/52885228-8a53ec80-313e-11e9-90b2-f04e8eec50b9.gif)

**Comment by jba** [comment]
OK, I think I understand your problem, but I don't understand your solution. You say replace the `go func()` line with an "unbounded pool of goroutines". Isn't that what that line already does? I mean, isn't the `go` keyword effectively starting a goroutine from an unbounded pool?

**Comment by dpcollins-google** [comment]
By default yes, if all goroutines end up blocking, it will.  But if the work per message is minimal and itself nonblocking, the user function will complete inline, reusing the goroutine instead of spawning a new one.  What the current behavior does is unconditionally create and tear down goroutines for every message.  If the time to start, run the user function and stop is greater than the time to trigger the start of a goroutine (aka the in thread time of go func(){}()) then the number of goroutines will grow, bounded only by the number of batches outstanding at the same time.

If you use an unbounded pool instead with goroutine reuse, and the time to trigger the start of a goroutine plus send a function on a channel is less than the time to receive on a channel and run the user function, the pool will only grow unbounded until the first goroutine starts, runs the user function, and starts receiving on the channel again.  It will then grow at a slower pace until there are enough goroutines to handle a batch before the next batch arrives.

In addition, by using an explicit pool construct, it enables the user to potentially (I would do this) replace the unbounded threadpool with a bounded threadpool, of NumCores workers looping as fast as possible, since they know the code they're running can't block forever.

I've put my code below, but once git codereview stops saying it needs formatting after I ran gofmt on it, ill submit a pr.

```golang
type WorkerPool interface {
	// Submit a new function to this pool
	Submit(func())
	// Wait until this pool has shut down.
	Stop() error
}

type unboundedWorkerPool struct {
	ctx      context.Context
	cancel   context.CancelFunc
	group    *errgroup.Group
	workChan chan func()
}

func (pool *unboundedWorkerPool) spawn() {
	startedChan := make(chan *types.Nil)
	pool.group.Go(func() error {
		startedChan <- nil
		for {
			select {
			case <-pool.ctx.Done():
				return nil
			case toRun := <-pool.workChan:
				toRun()
			}
		}
	})
	<-startedChan
}

func (pool *unboundedWorkerPool) Submit(toRun func()) {
	for {
		select {
		case pool.workChan <- toRun:
			return
		default:
			pool.spawn()
		}
	}
}

func (pool *unboundedWorkerPool) Stop() error {
	pool.cancel()
	return pool.group.Wait()
}

func MakeUnboundedWorkerPool(ctx context.Context, bufferSize int) WorkerPool {
	child, cancel := context.WithCancel(ctx)
	group, gctx := errgroup.WithContext(child)
	return &unboundedWorkerPool{
		ctx:      gctx,
		cancel:   cancel,
		group:    group,
		workChan: make(chan func(), bufferSize),
	}
}
```

**Comment by dpcollins-google** [comment]
https://code-review.googlesource.com/c/gocloud/+/38350

**Comment by jba** [comment]
@bcmills, does the argument for using a pool make sense to you? 

@jadekler, if we do switch, I would not do it as an option. No need to expose the complexity to the user. Just replace the current implementation.

**Comment by dpcollins-google** [comment]
@jba the decision to make the pool unbounded and growing immediately is primarily motivated by two things:

-Blocking a function on an asynchronous action is not an anti-pattern in go, so an arbitrary number of user callbacks may block.
-The existing implementation allows an arbitrary number of user callbacks to block.

However, I don't necessarily think this applies to all use cases; a re-publisher use case comes to mind, where the publish call is nearly nonblocking due to batching.  If the decision is made not to expose the pool concept for user injection, I would still like a parameter to control how many messages to build in backlog before launching a new goroutine (see bufferSize in MakeUnboundedWorkerPool in my cl).

**Comment by jba** [comment]
I've been playing around with this today. I do observe that having a pool, even an unbounded one, can bound memory usage. But coming up with the right pool logic is tricky. For instance, I couldn't find a good way to reduce the number of goroutines when the messages stopped coming.

I think this is worth further study. @dpcollins-google, maybe you can try your CL in production and see how it does. (Though I don't think we could take it as is because it doesn't shed goroutines during low load.)

**Comment by dpcollins-google** [comment]
@jba I updated this Sunday afternoon (~2 days ago) to shed goroutines under low load via a configurable keep alive duration.  I'd be interested to hear your thoughts.

**Comment by bcmills** [comment]
> [I]f the work per message is minimal and itself nonblocking, the user function will complete inline, reusing the goroutine instead of spawning a new one.

There is no fundamental reason why the behavior of the runtime with ephemeral goroutines should differ from that description. If you find a significant empirical difference in memory consumption, you should raise an issue against the runtime before contorting the code to work around it.

(See [Rethinking Classical Concurrency Patterns](https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view?usp=sharing) starting at slide 75.)

**Comment by dpcollins-google** [comment]
@bcmills 

Presuming that it is true that golang always reliably turns trivial goroutines into inline function calls or the equivalent, there is still the question of what is considered 'trivial'.  I'd hope that the following is:

```golang
func(i int) {
  return i + 1;
}
```

But I dont think this would be (although it would never in a realistic period of time block).

```golang
c := make(chan int, 1<<31)
func(i int) {
  c<-i
}
```

And I'm positive this would not be (which is effectively my use case) even though it will never block for long.

```golang
c := make(chan int, 100000)
go func() {
  sum := 0
  for {
  select {
    case val := <-c:
      sum = sum + val
    case <-reqChan:
      respChan<-sum
      sum = 0
  }
  }
}
func(m *Message) {
  derived := calcDerived(m) // pure mathematical operation
  c <- derived
  m.Ack()
}
```

When we're generating goroutines in a tight loop, especially since each of them has a ~4k overhead, when using a 1k payload, this results in significant memory overhead due to only goroutines.

I don't see it as an issue with the runtime that it can't inline this, but I don't think its an unreasonable statement to say that I'd expect the overhead of creating a nontrivial but short lived and efficient goroutine to be more than that of an existing goroutine reading a function object from a channel, and that the goroutine spawning would display significantly higher memory growth which is what I'm experiencing.

**Comment by bcmills** [comment]
> I don't think its an unreasonable statement to say that I'd expect the overhead of creating a nontrivial but short lived and efficient goroutine to be more than that of an existing goroutine reading a function object from a channel

On that point, we disagree: I would expect a short-lived goroutine to have very nearly the same cost as an existing goroutine reading and executing a function from a channel.

Specifically, if the goroutine does not end up blocking, the sequence should be essentially:
1. Check out a goroutine stack from a CPU-local pool of stacks.
2. Evaluate the goroutine body.
3. Return the goroutine stack to the thread-local pool.

Steps (1) and (3) might require the execution of a few extra instructions, but no data access outside of the local CPU cache. More importantly, the memory overhead should be O(CPUs) (for the pool of goroutine stacks) rather than O(goroutines), since each goroutine should return its stack to the pool when it retires.

**Comment by dpcollins-google** [comment]
@bcmills Is a write on a high concurrency channel with free space considered a block?  And what if a block does happen and the channel is contested for a short period of time; doesn't that mean that every goroutine being generated in the tight loop will start, hit the blocked channel, and in the meantime occupy 4k of space since it is still in progress?  Meanwhile the likelihood that the runtime will switch to the goroutine polling from the channel decreases rapidly as we are constantly spawning new ones (which I'd assume all get equal priority).

Furthermore, although a consumer of the google cloud pubsub library might have the ability to know that load is spiky and stick a buffered channel between the message receipt and the start of processing to decrease thrashing, if we unconditionally spawn a goroutine on every message this is not possible.  The api provided does not give you access to a queue of messages; it gives you a single callback.  While I could do no work other than pushing the messages to a make(chan Message*, 100000), this would eventually block causing the same issue.  What I would like to have as the result of this is the ability to delay goroutine spawning until there are N messages not yet being processed by a goroutine, and the ability to control N, which is what I have implemented, providing a buffer before the tight loop of goroutine spawning so the subscriber callback can make progress.

**Comment by bcmills** [comment]
> Is a write on a high concurrency channel with free space considered a block?

Presumably no.

> And what if a block does happen and the channel is contested for a short period of time; doesn't that mean that every goroutine being generated in the tight loop will start, hit the blocked channel, and in the meantime occupy 4k of space since it is still in progress?

Doesn't your “unbounded pool” proposal have exactly the same failure mode?

> Meanwhile the likelihood that the runtime will switch to the goroutine polling from the channel decreases rapidly as we are constantly spawning new ones (which I'd assume all get equal priority).

When a channel operation unblocks a goroutine, the goroutine scheduler prefers to switch to that goroutine. With typical usage patterns, that tends to reduce the number of live goroutines under load. (Goroutines are not OS threads, so thread-based assumptions like “equal priority” do not necessarily hold.)

**Comment by dpcollins-google** [comment]
> Doesn't your “unbounded pool” proposal have exactly the same failure mode?

If you set the buffer size and time for a goroutine without doing anything to 0, yes, because it becomes identical to spawning one goroutine per message.  If you have a buffer of the size of a batch or larger though, it becomes possible to spawn 0 goroutines for each batch that comes in once you have enough already outstanding to handle the load before the next batch arrives.  Because load comes in in discrete batches, in the current state, you're guaranteed to spawn ten thousand goroutines when every batch message of 10MB of 1KB messages arrives, before there is any time to handle any of that load.

In my case, user memory is less than half of the machine's available memory, but my process is getting killed by running out of memory- presumably due to outstanding goroutine overhead.  If you have a more idiomatic go suggestion to resolve high throughput small message subscribers running out of memory, I would greatly appreciate it.

**Comment by jba** [comment]
Sorry I haven't been keeping up on this; I was on vacation until today.

@dpcollins-google, how much time does it take you to process one message? I want to try to reproduce what you're doing as closely as possible.

> I updated ... to shed goroutines under low load via a configurable keep alive duration. I'd be interested to hear your thoughts.

Have you benchmarked? I found that timers don't scale as well as goroutines. If you have one timer per goroutine and millions of goroutines, performance will suffer.

**Comment by dpcollins-google** [comment]
@jba: all good, thanks for looking at this.

The full pr: https://github.com/GoogleCloudPlatform/pubsub/pull/192/files
The file calling Receive: https://github.com/GoogleCloudPlatform/pubsub/pull/192/files#diff-5bfee1484995548e998b3cc46407293b
Metrics tracker (blocking function called by receive): https://github.com/GoogleCloudPlatform/pubsub/pull/192/files#diff-94f71f092987628d6f9bf043ac85972d

As far as benchmarking, I'd appreciate any guidance you have.  It doesn't seem like the go memory profiler captures number of goroutines.  Do any of the pprof profilers output this metric?  Otherwise perhaps a waitgroup with printing would do.

In what way don't they scale as well?  Does a timer have the same memory overhead as a blocked goroutine or were you referring to CPU overhead?  In which case, I haven't been concerned by the cpu overhead of millions of blocked goroutines as much as the memory overhead causing crashes when user memory is seemingly safe.

I'd like to clarify that my concern is not so much whether I can make my use case work- I can, as you see, have a manual messages outstanding limit holding messages outstanding down to <50% of machine memory.  But I'm concerned that it seems to perform arbitrarily bad (by crashing when it should be memory limited by flow control) in what I see as a high throughput but not unreasonable use case.

**Comment by jba** [comment]
Ah, you are working on the pubsub load test and are not an actual user! That changes things. For one thing, it lowers the priority of this issue (in my way of thinking, at least; Jean is the owner). We don't have an actual customer running into this problem. That said, it clearly _could_ end up being a problem for someone down the road, so I'm not saying we want to ignore it. I would also be concerned if this caused Go's benchmarks to look significantly worse that Java's, although even there I would want to know if anyone really needed that level of throughput in practice.

I looked at your code. Your receive callback doesn't do much; I measured a simulated version of it as taking a little over a millisecond. I rounded up to 2ms and ran
```
    for {
        go func() {
            time.Sleep(2*time.Millisecond)
        }()
        time.Sleep(0) // something like this is needed so the goroutine yields
    }
}
```
on my 8-CPU machine. Until I ran that loop in 7 separate goroutines, I couldn't get memory to grow without bound. In other words, when the goroutines finish quickly, it's hard to generate them fast enough to run out of RAM.  

I'm guessing that either `request.CpuScaling` is greater than 1, or maybe that the channel inside `metricsTracker.Put` is filling up, delaying the goroutine.

> It doesn't seem like the go memory profiler captures number of goroutines.

You can call `runtime.NumGoroutine`, although I think of more direct interest would be `MemStats.StackInuse`. My little program displays that number and `HeapAlloc` at 1-second intervals as I simulate receiving messages in a tight loop.

> In what way don't they scale as well?

I don't know the internals, but when I added a `time.After` to each goroutine so it could die when idle, my throughput plummeted.

**Comment by meredithslota** [comment]
It's unclear to me what else needs to be done to close out this issue — it seems like the conversation slowed down and some interim commits happened to update loadtest framework, but then ??? 

I'm tempted to close this and revisit if needed in a new issue. :) @hongalex or @kamalaboulhosn feel free to reopen if you think there's an immediate need.
